torch/csrc/jit/passes/device_type_analysis.cpp

```diff
@@ -249,10 +255,30 @@ struct DeviceTypePropagationPass {
     if (!op) {
       return;
     }
+    auto prop_fn = device_prop_registry_->find(*op);
+    if (prop_fn) {
+      PropRule rule = *prop_fn;
+      changed_ |= rule(n);
+      return;
+    }
     changed_ |= defaultDeviceProp(n);
   }
 
+  void buildRuleRegistry() {
+    // building a registry for all of the custom Device Type rules
+    static const OperatorMap<PropRule> temp_registry{
+        {"aten::cpu(Tensor self) -> Tensor",
+         setReturnstoDeviceRule(DeviceType::CPU)},
+        {"aten::cuda(Tensor self) -> Tensor",
+         setReturnstoDeviceRule(DeviceType::CUDA)},
+        {"aten::to_mkldnn(Tensor self, ScalarType? dtype) -> Tensor",
+         setReturnstoDeviceRule(DeviceType::MKLDNN)},
+    };
+    device_prop_registry_ =
```

Comment:
is this making a copy, otherwise we might get double free if we instantiate more than one `DeviceTypePropagationPass` since both will try to `delete` the same pointer?
Is there a reason we could store a reference to `temp_registry` or better yet just use it as a global singleton?

```
OperatorMap<PropRule>& device_prop_registry_;
```

Reply:
What is your recommended way of making a global singleton?